// @aws-amplify/deployed-backend-client/lib/backend_output_client_factory.js
import { DefaultBackendOutputClient } from './backend_output_client.js';
import { CloudFormationClient } from '@aws-sdk/client-cloudformation';
import { AmplifyClient } from '@aws-sdk/client-amplify';
export var BackendOutputClientErrorType;
(function (BackendOutputClientErrorType) {
    BackendOutputClientErrorType["METADATA_RETRIEVAL_ERROR"] = "MetadataRetrievalError";
    BackendOutputClientErrorType["NO_OUTPUTS_FOUND"] = "NoOutputsFound";
    BackendOutputClientErrorType["DEPLOYMENT_IN_PROGRESS"] = "DeploymentInProgress";
    BackendOutputClientErrorType["NO_STACK_FOUND"] = "NoStackFound";
    BackendOutputClientErrorType["CREDENTIALS_ERROR"] = "CredentialsError";
    BackendOutputClientErrorType["ACCESS_DENIED"] = "AccessDenied";
})(BackendOutputClientErrorType || (BackendOutputClientErrorType = {}));
/**
 * Error type for BackendOutputClientError
 */
export class BackendOutputClientError extends Error {
    code;
    /**
     * Constructor for BackendOutputClientError
     */
    constructor(code, message, options) {
        super(message, options);
        this.code = code;
    }
    /**
     * This function is a type predicate for BackendOutputClientError.
     * See https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates.
     *
     * Checks if error is an BackendOutputClientError by inspecting if required properties are set.
     * This is recommended instead of instanceof operator.
     * The instance of operator does not work as expected if BackendOutputClientError class is loaded
     * from multiple sources, for example when package manager decides to not de-duplicate dependencies.
     * See https://github.com/nodejs/node/issues/17943.
     */
    static isBackendOutputClientError = (error) => {
        return (error instanceof Error &&
            'code' in error &&
            typeof error.code === 'string' &&
            Object.values(BackendOutputClientErrorType).includes(error.code) &&
            typeof error.message === 'string');
    };
}
/**
 * Factory to create a backend metadata reader
 */
export class BackendOutputClientFactory {
    /**
     * Returns a single instance of BackendOutputClient
     */
    static getInstance = (awsClientProvider) => {
        return new DefaultBackendOutputClient(awsClientProvider?.getCloudFormationClient() ??
            new CloudFormationClient(), awsClientProvider?.getAmplifyClient() ?? new AmplifyClient());
    };
}
