// @aws-amplify/client-config/lib/client-config-types/client_config.js
// Versions of new unified config schemas
import * as clientConfigTypesV1 from '../client-config-schema/client_config_v1.js';
/* eslint-disable @typescript-eslint/naming-convention */
import * as clientConfigTypesV1_1 from '../client-config-schema/client_config_v1.1.js';
import * as clientConfigTypesV1_2 from '../client-config-schema/client_config_v1.2.js';
import * as clientConfigTypesV1_3 from '../client-config-schema/client_config_v1.3.js';
export { clientConfigTypesV1, clientConfigTypesV1_1, clientConfigTypesV1_2, clientConfigTypesV1_3, };
export var ClientConfigVersionOption;
(function (ClientConfigVersionOption) {
    ClientConfigVersionOption["V0"] = "0";
    ClientConfigVersionOption["V1"] = "1";
    ClientConfigVersionOption["V1_1"] = "1.1";
    ClientConfigVersionOption["V1_2"] = "1.2";
    ClientConfigVersionOption["V1_3"] = "1.3";
})(ClientConfigVersionOption || (ClientConfigVersionOption = {}));
// Client config version that is generated by default if customers didn't specify one
export const DEFAULT_CLIENT_CONFIG_VERSION = ClientConfigVersionOption.V1_3;
export var ClientConfigFormat;
(function (ClientConfigFormat) {
    ClientConfigFormat["MJS"] = "mjs";
    ClientConfigFormat["JSON"] = "json";
    ClientConfigFormat["JSON_MOBILE"] = "json-mobile";
    ClientConfigFormat["TS"] = "ts";
    ClientConfigFormat["DART"] = "dart";
})(ClientConfigFormat || (ClientConfigFormat = {}));
export var ClientConfigFileBaseName;
(function (ClientConfigFileBaseName) {
    ClientConfigFileBaseName["LEGACY"] = "amplifyconfiguration";
    ClientConfigFileBaseName["DEFAULT"] = "amplify_outputs";
})(ClientConfigFileBaseName || (ClientConfigFileBaseName = {}));
